src/Build.UnitTests/BackEnd/TaskBuilder_Tests.cs

```diff
@@ -168,8 +170,18 @@ public void CanceledTasksDoNotLogMSB4181()
                 BuildRequestData data = new BuildRequestData(project.CreateProjectInstance(), new string[] { "test" }, collection.HostServices);
                 manager.BeginBuild(_parameters);
                 BuildSubmission asyncResult = manager.PendBuildRequest(data);
+                string unescapedSleepCommand = sleepCommand.Replace("&quot;", "\"").Replace("&gt;", ">");
+                Func<bool> isSleepCommandExecuted = () => logger.AllBuildEvents.Any(a => unescapedSleepCommand.Equals(a.Message));
```

Comment:
While accomplishing the purpose, it feels superfluous to iterate all the log events over and over again (I guess it risk getting the collection modified exception as well).

It should be easy to check only the new events (if any) on each try

Reply:
It seems [`AdditionalHandlers`](https://github.com/dotnet/msbuild/blob/f4fa6bde775a3f7cbb2bb90a349ee5fc759114f3/src/Shared/UnitTests/MockLogger.cs#L248C53-L248C71) can be leveraged here - by registering a handler that would set some waithandle upon receiving the required log entry